clusterloader2/pkg/util/util.go

```diff
@@ -24,50 +24,102 @@ import (
 	"time"
 )
 
+var errKeyNotFound = fmt.Errorf("key not found")
+
 // GetString tries to return value from map cast to string type. If value doesn't exist, error is returned.
 func GetString(dict map[string]interface{}, key string) (string, error) {
-	return getString(dict, key, "", fmt.Errorf("key %s not found", key))
+	value, err := getString(dict, key)
+	if err == errKeyNotFound {
+		return value, fmt.Errorf("key %s not found", key)
+	}
+	return value, err
 }
 
 // GetInt tries to return value from map cast to int type. If value doesn't exist, error is returned.
 func GetInt(dict map[string]interface{}, key string) (int, error) {
-	return getInt(dict, key, 0, fmt.Errorf("key %s not found", key))
+	value, err := getInt(dict, key)
+	if err == errKeyNotFound {
+		return value, fmt.Errorf("key %s not found", key)
+	}
+	return value, err
 }
 
 // GetFloat64 tries to return value from map cast to float64 type. If value doesn't exist, error is returned.
 func GetFloat64(dict map[string]interface{}, key string) (float64, error) {
-	return getFloat64(dict, key, 0, fmt.Errorf("key %s not found", key))
+	value, err := getFloat64(dict, key)
+	if err == errKeyNotFound {
+		return value, fmt.Errorf("key %s not found", key)
+	}
+	return value, err
 }
 
 // GetDuration tries to return value from map cast to duration type. If value doesn't exist, error is returned.
 func GetDuration(dict map[string]interface{}, key string) (time.Duration, error) {
-	return getDuration(dict, key, 0, fmt.Errorf("key %s not found", key))
+	value, err := getDuration(dict, key)
+	if err == errKeyNotFound {
+		return value, fmt.Errorf("key %s not found", key)
+	}
+	return value, err
+}
+
+// GetBool tries to return value from map cast to bool type. If value doesn't exist, error is returned.
+func GetBool(dict map[string]interface{}, key string) (bool, error) {
+	value, err := getBool(dict, key)
+	if err == errKeyNotFound {
+		return value, fmt.Errorf("key %s not found", key)
+	}
+	return value, err
 }
 
 // GetStringOrDefault tries to return value from map cast to string type. If value doesn't exist default value is used.
 func GetStringOrDefault(dict map[string]interface{}, key string, defaultValue string) (string, error) {
-	return getString(dict, key, defaultValue, nil)
+	value, err := getString(dict, key)
```

Comment:
What are the usecases for "Get*OrDefault" - do we even use non-zero value for defaults?

Reply:
> // GetStringOrDefault tries to return value from map cast to string type. If value doesn't exist default value is used.

The assumption is that value can be optional - will not exists in the map. In this case you want to use default value (not necessary a zero value) specified for certain use case.

Use case:
In metric we can enable option to gather data only from master, so we use:
```util.GetBoolOrDefault(config.Params, "masterOnly", false)```
which assumes that user can specify value or default (false) will be used.

Reply:
There are additional two in `pkg/measurement/common/scheduler_latency.go`:
```host, err := util.GetStringOrDefault(config.Params, "host", "")```
```masterName, err := util.GetStringOrDefault(config.Params, "masterName", "")```

Four more in `pkg/measurement/common/wait_for_controlled_pods` (and also in `wait_for_pods`):
```namespace, err := util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)```
```labelSelector, err := util.GetStringOrDefault(config.Params, "labelSelector", "")```
```fieldSelector, err := util.GetStringOrDefault(config.Params, "fieldSelector", "")```
```timeout, err := util.GetDurationOrDefault(config.Params, "timeout", defaultWaitForRCPodsTimeout)```

There are more use cases in metric I'm currently trying to add to clusterloader.
Assuming that measurements are the main part that will be developed, number of use cases will grow.

Reply:
Most of these aren't good examples, because they are using zero-value ("" is string zero-value).
NamespaceAll also turns out to be empty string :)

So timeout is the only case where you need it.

Reply:
> Most of these aren't good examples, because they are using zero-value ("" is string zero-value).
NamespaceAll also turns out to be empty string :)

But still I think that those are the places when you want to use already defined function instead of implementing error handling on your own. In these cases, you still want to use zero value **only if** value is not provided. Plus, having a way to define default value is convenient.

Reply:
After offline discussion we agreed to make following changes:
- Making the ErrKeyNotFound globally accessible.
- Added function to check whether a given error is ErrKeyNotFound type.